codec/reflectcodec/type_codec.go

```diff
@@ -520,6 +594,42 @@ func (c *genericCodec) unmarshal(p *wrappers.Packer, value reflect.Value, maxSli
 		// Assign to the top-level struct's member
 		value.Set(v)
 		return nil
+	case reflect.Map:
+		numElts32 := p.UnpackInt()
+		if p.Err != nil {
+			return fmt.Errorf("couldn't unmarshal map: %w", p.Err)
+		}
+		if numElts32 > c.maxSliceLen {
+			return fmt.Errorf("%w; map length, %d, exceeds maximum length, %d",
+				codec.ErrMaxSliceLenExceeded,
+				numElts32,
+				c.maxSliceLen,
+			)
+		}
+
+		numElts := int(numElts32)
+		valueTypeReflection := value.Type()
+		mapKeyType := valueTypeReflection.Key()
+		mapValueType := valueTypeReflection.Elem()
+
+		value.Set(reflect.MakeMapWithSize(valueTypeReflection, numElts))
+
+		for i := 0; i < numElts; i++ {
+			mapKey := reflect.New(mapKeyType).Elem()
+			mapValue := reflect.New(mapValueType).Elem()
+
+			if err := c.unmarshal(p, mapKey, c.maxSliceLen); err != nil {
+				return fmt.Errorf("couldn't unmarshal map key (%s): %w", mapKeyType, err)
+			}
+
+			if err := c.unmarshal(p, mapValue, c.maxSliceLen); err != nil {
+				return fmt.Errorf("couldn't unmarshal map value for key %s: %w", mapKey, err)
+			}
+			value.SetMapIndex(mapKey, mapValue)
+		}
```

Comment:
Should we error if the keys are not sorted?

Reply:
@aaronbuchwald In my opinion we shouldn't. Keys may or may not be sorted. It has no meaning in this context (when deserializing).

I think the sorting is a nice property when serializing, but adds zero benefit deserializing sorted or unsorted keys/value pairs.

Reply:
I also think we should enforce the sorting on deserialization to ensure that the message is well-formed

Reply:
I'll make it happen.

Reply:
Done in 07ea59217. Does it look good @danlaine @aaronbuchwald ?